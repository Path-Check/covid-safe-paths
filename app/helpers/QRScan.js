import { NativeModules } from 'react-native';

export const isValidCoordinates = (latitude, longitude) => {
  const isValid = num => {
    const regex = /^(-?\d+\.?\d*|\.\d+)$/;
    return typeof num === 'string' && regex.test(num);
  };
  return isValid(latitude) && isValid(longitude);
};

export const parseQRCodeUrl = url => {
  if (typeof url !== 'string' || url.search('safepaths://qr/') !== 0) {
    return {};
  }
  const split1 = url.split('/qr/');
  const split2 = split1 && split1.length === 2 && split1[1].split('/');
  const latitude = (split2 && split2.length === 2 && split2[0]) || undefined;
  const longitude = (split2 && split2.length === 2 && split2[1]) || undefined;
  return { latitude, longitude };
};

export const saveLocation = async location => {
  await NativeModules.SecureStorageManager.importGoogleLocations([location]);
};
