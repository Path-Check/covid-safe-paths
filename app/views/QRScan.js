import React, { Component } from 'react';
import {
  BackHandler,
  Dimensions,
  ImageBackground,
  StatusBar,
  StyleSheet,
  View,
} from 'react-native';
import { SvgXml } from 'react-native-svg';

import BackgroundImage from './../assets/images/launchScreenBackground.png';
import StateAtRisk from './../assets/svgs/stateAtRisk';
import StateNoContact from './../assets/svgs/stateNoContact';
import { isPlatformAndroid } from './../Util';
import ButtonWrapper from '../components/ButtonWrapper';
import { Typography } from '../components/Typography';
import Colors from '../constants/colors';
import fontFamily from '../constants/fonts';
import languages from '../locales/languages'; // TODO internationalization
import LocationServices from '../services/LocationService';

const StateEnum = {
  SCAN_SUCCESS: 0,
  SCAN_FAIL: 1,
};

const StateIcon = ({ status, size }) => {
  let icon;
  switch (status) {
    case StateEnum.SCAN_SUCCESS:
      icon = StateNoContact;
      break;
    case StateEnum.SCAN_FAIL:
      icon = StateAtRisk;
      break;
  }
  return (
    <SvgXml xml={icon} width={size ? size : 80} height={size ? size : 80} />
  );
};

const height = Dimensions.get('window').height;

class QRScan extends Component {
  constructor(props) {
    super(props);

    if (isPlatformAndroid()) {
      StatusBar.setBackgroundColor(Colors.TRANSPARENT);
      StatusBar.setBarStyle('light-content');
      StatusBar.setTranslucent(true);
    }

    this.state = {
      currentState: StateEnum.SCAN_SUCCESS,
    };
  }

  componentDidMount() {
    BackHandler.addEventListener('hardwareBackPress', this.handleBackPress);
    this.unsubscribeFocus = this.props.navigation.addListener('focus', () => {
      this.processRouteParams();
    });
    this.unsubscribeState = this.props.navigation.addListener('state', () => {
      this.processRouteParams();
    });
  }

  componentWillUnmount() {
    BackHandler.removeEventListener('hardwareBackPress', this.handleBackPress);
    this.unsubscribeFocus();
    this.unsubscribeState();
  }

  processRouteParams() {
    const {
      latitudeInteger,
      latitudeFraction,
      longitudeInteger,
      longitudeFraction,
    } = this.props.route.params;
    this.saveCoordinates(
      latitudeInteger,
      latitudeFraction,
      longitudeInteger,
      longitudeFraction,
    );
  }

  saveCoordinates(
    latitudeInteger,
    latitudeFraction,
    longitudeInteger,
    longitudeFraction,
  ) {
    if (
      this.isValidCoordinates(
        latitudeInteger,
        latitudeFraction,
        longitudeInteger,
        longitudeFraction,
      )
    ) {
      const latitudeParsed = Number(`${latitudeInteger}.${latitudeFraction}`);
      const longitudeParsed = Number(
        `${longitudeInteger}.${longitudeFraction}`,
      );
      LocationServices.saveLocation({
        latitude: latitudeParsed,
        longitude: longitudeParsed,
        time: Date.now(),
      });
      this.setState({ currentState: StateEnum.SCAN_SUCCESS });
    } else {
      this.setState({ currentState: StateEnum.SCAN_FAIL });
    }
  }

  isValidCoordinates(
    latitudeInteger,
    latitudeFraction,
    longitudeInteger,
    longitudeFraction,
  ) {
    const isValidInteger = num => {
      const regex = /^\d+?$/;
      return typeof num === 'string' && regex.test(num);
    };
    return (
      isValidInteger(latitudeInteger) &&
      isValidInteger(latitudeFraction) &&
      isValidInteger(longitudeInteger) &&
      isValidInteger(longitudeFraction)
    );
  }

  handleBackPress = () => {
    BackHandler.exitApp(); // works best when the goBack is async
    return true;
  };

  getBackground() {
    return BackgroundImage;
  }

  getMainText() {
    switch (this.state.currentState) {
      case StateEnum.SCAN_SUCCESS:
        return (
          <Typography style={styles.mainTextBelow}>
            {languages.t('label.qr_successful_title')}
          </Typography>
        );
      case StateEnum.SCAN_FAIL:
        return (
          <Typography style={styles.mainTextBelow}>
            {languages.t('label.qr_fail_title')}
          </Typography>
        );
    }
  }

  getSubText() {
    switch (this.state.currentState) {
      case StateEnum.SCAN_SUCCESS:
        return languages.t('label.qr_successful_subtitle');
      case StateEnum.SCAN_FAIL:
        return languages.t('label.qr_fail_subtitle');
    }
  }

  render() {
    return (
      <ImageBackground
        source={this.getBackground()}
        style={styles.backgroundImage}>
        <StatusBar
          barStyle='light-content'
          backgroundColor='transparent'
          translucent
        />
        <View style={styles.iconContainer}>
          <StateIcon size={height} status={this.state.currentState} />
        </View>
        <View style={styles.mainContainer}>
          <View style={styles.content}>
            {this.getMainText()}
            <Typography style={styles.subheaderText}>
              {this.getSubText()}
            </Typography>
            <View style={styles.buttonContainer}>
              <ButtonWrapper
                title='Home'
                onPress={() => {
                  this.props.navigation.navigate('LocationTrackingScreen', {});
                }}
                buttonColor={Colors.BLUE_BUTTON}
                bgColor={Colors.WHITE}
              />
            </View>
          </View>
        </View>
      </ImageBackground>
    );
  }
}

const ICON_GAP = 80;

const styles = StyleSheet.create({
  backgroundImage: {
    width: '100%',
    height: '100%',
    resizeMode: 'cover',
    flex: 1,
    justifyContent: 'flex-end',
  },
  mainContainer: {
    position: 'absolute',
    top: '-10%',
    left: 0,
    right: 0,
    height: '90%',
    paddingHorizontal: '12%',
    paddingBottom: 12,
  },
  content: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingTop: ICON_GAP,
  },
  iconContainer: {
    position: 'absolute',
    resizeMode: 'contain',
    height: '100%',
    top: '-13%',
    left: 0,
    right: 0,
    alignItems: 'center',
  },
  buttonContainer: {
    top: 24,
  },
  mainTextBelow: {
    textAlign: 'center',
    lineHeight: 34,
    color: Colors.WHITE,
    fontSize: 26,
    fontFamily: fontFamily.primaryMedium,
    marginBottom: 24,
  },
  subheaderText: {
    marginBottom: 24,
    textAlign: 'center',
    lineHeight: 24.5,
    color: Colors.WHITE,
    fontSize: 18,
    fontFamily: fontFamily.primaryRegular,
  },
});

export default QRScan;
