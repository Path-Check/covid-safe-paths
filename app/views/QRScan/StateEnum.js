export const StateEnum = {
  DEFAULT: 0,
  SCAN_SUCCESS: 1,
  SCAN_FAIL: 2,
  SCAN_IN_PROGRESS: 3,
};
