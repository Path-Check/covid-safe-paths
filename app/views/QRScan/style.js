import { StyleSheet } from 'react-native';

import Colors from '../../constants/colors';
import fontFamily from '../../constants/fonts';

const ICON_GAP = 80;

export const styles = StyleSheet.create({
  backgroundImage: {
    width: '100%',
    height: '100%',
    resizeMode: 'cover',
    flex: 1,
    justifyContent: 'flex-end',
  },
  mainContainer: {
    position: 'absolute',
    top: '-10%',
    left: 0,
    right: 0,
    height: '90%',
    paddingHorizontal: '12%',
    paddingBottom: 12,
  },
  qrScanContainer: {
    height: '100%',
    paddingBottom: 10,
  },
  content: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingTop: ICON_GAP,
  },
  iconContainer: {
    position: 'absolute',
    resizeMode: 'contain',
    height: '100%',
    top: '-13%',
    left: 0,
    right: 0,
    alignItems: 'center',
  },
  buttonContainer: {
    top: 24,
  },
  mainTextBelow: {
    lineHeight: 34,
    fontSize: 26,
    marginBottom: 24,
  },
  subheaderText: {
    marginBottom: 24,
    textAlign: 'center',
  },
});
