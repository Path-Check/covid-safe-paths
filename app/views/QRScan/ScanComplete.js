import React from 'react';
import { useTranslation } from 'react-i18next';
import { Dimensions, View } from 'react-native';
import { SvgXml } from 'react-native-svg';

import StateAtRisk from '../../assets/svgs/stateAtRisk';
import StateNoContact from '../../assets/svgs/stateNoContact';
import { Button } from '../../components/Button';
import { Typography } from '../../components/Typography';
import { StateEnum } from './StateEnum';
import { styles } from './style';

const height = Dimensions.get('window').height;

const StateIcon = ({ status, size }) => {
  let icon;
  switch (status) {
    case StateEnum.DEFAULT:
      icon = null;
      break;
    case StateEnum.SCAN_SUCCESS:
      icon = StateNoContact;
      break;
    case StateEnum.SCAN_FAIL:
      icon = StateAtRisk;
      break;
  }
  return <SvgXml xml={icon} width={size || 80} height={size || 80} />;
};

export const ScanComplete = ({ currentState, onExit }) => {
  const { t } = useTranslation();

  const getMainText = () => {
    switch (currentState) {
      case StateEnum.SCAN_SUCCESS:
        return (
          <Typography style={styles.mainTextBelow} use='headline2'>
            {t('qr.successful_title')}
          </Typography>
        );
      case StateEnum.SCAN_FAIL:
        return (
          <Typography style={styles.mainTextBelow} use='headline2'>
            {t('qr.fail_title')}
          </Typography>
        );
    }
  };

  const getSubText = () => {
    switch (currentState) {
      case StateEnum.SCAN_SUCCESS:
        return t('qr.successful_subtitle');
      case StateEnum.SCAN_FAIL:
        return t('qr.fail_subtitle');
    }
  };

  return (
    <View style={styles.ScanCompleteContainer}>
      <View style={styles.iconContainer}>
        <StateIcon size={height} status={currentState} />
      </View>
      <View style={styles.mainContainer}>
        <View style={styles.content}>
          {getMainText()}
          <Typography style={styles.subheaderText} use='body1'>
            {getSubText()}
          </Typography>
          <View style={styles.buttonContainer}>
            <Button label={t('qr.exit')} onPress={onExit} />
          </View>
        </View>
      </View>
    </View>
  );
};
