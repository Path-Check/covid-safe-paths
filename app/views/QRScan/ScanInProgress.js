import React from 'react';
import { useTranslation } from 'react-i18next';
import { View } from 'react-native';
import { RNCamera } from 'react-native-camera';
import QRCodeScanner from 'react-native-qrcode-scanner';

import { Button } from '../../components/Button';
import { styles } from './style';

export const ScanInProgress = ({ onScanCancel, onScanSuccess }) => {
  const { t } = useTranslation();

  return (
    <View style={styles.qrScanContainer}>
      <QRCodeScanner
        onRead={onScanSuccess}
        flashMode={RNCamera.Constants.FlashMode.off}
      />
      <Button
        label={t('qr.scan_cancel')}
        style={styles.qrCancelButton}
        onPress={onScanCancel}
      />
    </View>
  );
};
