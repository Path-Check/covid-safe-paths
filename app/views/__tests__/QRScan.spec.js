import 'react-native';

import { render, wait } from '@testing-library/react-native';
import React from 'react';

import QRScan from '../QRScan';

it('renders correctly', async () => {
  const createTestProps = props => ({
    navigation: {
      addListener: jest.fn(),
    },
    ...props,
  });
  const props = createTestProps({});
  const { asJSON } = render(<QRScan {...props} />);
  await wait();

  expect(asJSON()).toMatchSnapshot();
});
