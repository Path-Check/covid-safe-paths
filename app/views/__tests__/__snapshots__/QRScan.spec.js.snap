// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    accessibilityIgnoresInvertColors={true}
    style={
      Object {
        "flex": 1,
        "height": "100%",
        "justifyContent": "flex-end",
        "resizeMode": "cover",
        "width": "100%",
      }
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../app/assets/images/launchScreenBackground.png",
        }
      }
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          },
          Object {
            "height": "100%",
            "width": "100%",
          },
          undefined,
        ]
      }
    />
  </View>
</View>
`;
